src/fhirpath.js

```diff
@@ -505,7 +523,13 @@ engine.FunctionInvocation = function(ctx, parentData, node) {
   const fnName = args[0];
   args.shift();
   var rawParams = args && args[0] && args[0].children;
-  return doInvoke(ctx, fnName, parentData, rawParams);
+  if (fnName == 'defineVariable') {
+    return doInvoke(ctx, fnName, parentData, rawParams);
+  }
+  let ctxForArgs = { ...ctx };
+  if (ctx.definedVars)
+    ctxForArgs.definedVars = { ... ctx.definedVars };
+  return doInvoke(ctxForArgs, fnName, parentData, rawParams);
```

Comment:
`ctxForArgs` is used not only for arguments, but also for the function. So, calling each function will create a new context, but for `defineVariable` you made an exception, that is why my unit test failed. I think that a copy of the context needs to be created not for the function but for the argument (in the `makeParam()` function).